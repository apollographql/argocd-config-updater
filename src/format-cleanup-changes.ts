import { dirname } from 'node:path';

export interface CleanupChange {
  prNumber: number;
  prTitle: string;
  prURL: string;
  filename: string;
  environment: string;
  closedAt: string | null;
}

export function formatCleanupChanges(changes: CleanupChange[]): string {
  if (!changes.length) return '';

  // Group by environment, then by app directory
  const grouped = new Map<string, Map<string, CleanupChange[]>>();

  for (const change of changes) {
    const appDir = dirname(change.filename);
    const env = change.environment;

    let envGroup = grouped.get(env);
    if (!envGroup) {
      envGroup = new Map();
      grouped.set(env, envGroup);
    }

    let appChanges = envGroup.get(appDir);
    if (!appChanges) {
      appChanges = [];
      envGroup.set(appDir, appChanges);
    }

    appChanges.push(change);
  }

  const lines: string[] = [];
  const sortedEnvs = [...grouped.entries()].sort(([a], [b]) =>
    a.localeCompare(b),
  );

  for (const [environment, appGroups] of sortedEnvs) {
    lines.push(`## ${environment}`);
    lines.push('');

    const sortedApps = [...appGroups.entries()].sort(([a], [b]) =>
      a.localeCompare(b),
    );

    for (const [appDir, appChanges] of sortedApps) {
      lines.push(`- ${appDir}`);

      const sortedChanges = appChanges.sort((a, b) => a.prNumber - b.prNumber);
      for (const change of sortedChanges) {
        const closedDateStr = change.closedAt
          ? ` (closed ${new Date(change.closedAt).toLocaleDateString('en-US', {
              month: 'short',
              day: 'numeric',
              year: 'numeric',
            })})`
          : '';

        lines.push(
          `- PR [#${change.prNumber}](${change.prURL}): ${change.prTitle}${closedDateStr}`,
        );
      }
      lines.push('');
    }
  }

  const header = `# Found ${changes.length} closed PR${changes.length === 1 ? '' : 's'}\n\n`;
  return header + lines.join('\n').trim();
}
