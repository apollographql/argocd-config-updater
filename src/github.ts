import * as core from '@actions/core';
import { getOctokit } from '@actions/github';
import { LRUCache } from 'lru-cache';

export interface ResolveRefToSHAOptions {
  repoURL: string;
  ref: string;
}

export interface GetTreeSHAForPathOptions {
  repoURL: string;
  commitSHA: string;
  path: string;
}

export interface CompareCommitsOptions {
  repoURL: string;
  baseCommitSHA: string;
  headCommitSHA: string;
}

export interface CompareCommitsResult {
  commits: {
    commitSHA: string;
    message: string;
    author: string | null;
    commitUrl: string;
  }[];
}

export interface GitHubClient {
  resolveRefToSHA(options: ResolveRefToSHAOptions): Promise<string>;
  getTreeSHAForPath(options: GetTreeSHAForPathOptions): Promise<string | null>;
  compareCommits(
    options: CompareCommitsOptions,
  ): Promise<CompareCommitsResult | null>;
}

interface OwnerAndRepo {
  owner: string;
  repo: string;
}

function parseRepoURL(repoURL: string): OwnerAndRepo {
  const m = repoURL.match(/\bgithub\.com\/([\w.-]+)\/([\w.-]+?)(?:\.git|\/)?$/);
  if (!m) {
    throw Error(`Can only track GitHub repoURLs, not ${repoURL}`);
  }
  return { owner: m[1], repo: m[2] };
}

interface AllTreesForCommit {
  pathToTreeSHA: Map<string, string>;
  // If true, GitHub's response to the recursive tree fetch was truncated, so on
  // cache miss we fall back to the getContent API.
  truncated: boolean;
}
export class OctokitGitHubClient {
  apiCalls = new Map<string, number>();
  constructor(private octokit: ReturnType<typeof getOctokit>) {}

  async compareCommits({
    repoURL,
    baseCommitSHA,
    headCommitSHA,
  }: CompareCommitsOptions): Promise<CompareCommitsResult | null> {
    const baseIsCommitSHA = baseCommitSHA.match(/([0-9a-fA-F]+)$/);
    const headIsCommitSHA = headCommitSHA.match(/([0-9a-fA-F]+)$/);
    if (!baseIsCommitSHA || !headIsCommitSHA) return null;

    const { owner, repo } = parseRepoURL(repoURL);

    // Include '^' to be inclusive of the head commit.
    const basehead = `${baseCommitSHA}...${headCommitSHA}^`;
    this.logAPICall('repos.compareCommits', `${owner}/${repo} ${basehead}`);
    try {
      const result = (
        await this.octokit.rest.repos.compareCommitsWithBasehead({
          owner,
          repo,
          basehead,
        })
      ).data.commits;
      return {
        commits: result.map((commit) => ({
          commitSHA: commit.sha,
          message: commit.commit.message,
          author: commit.author?.name ?? commit.commit.author?.email ?? null,
          commitUrl: commit.html_url,
        })),
      };
    } catch (error: unknown) {
      core.info(`[GH API] Error in compareCommits, ignoring`);
      return null;
    }
  }

  private logAPICall(name: string, description: string): void {
    core.info(`[GH API]${name} ${description}`);
    this.apiCalls.set(name, (this.apiCalls.get(name) ?? 0) + 1);
  }

  // The cache key is JSON-ification of `{ repoURL, commitSHA }`.
  private allTreesForCommitCache = new LRUCache<
    string,
    AllTreesForCommit,
    { repoURL: string; commitSHA: string }
  >({
    max: 1024,
    // We ignore the key itself and treat the context as the key
    // (the key is just the JSON-ification of context).
    fetchMethod: async (_key, _staleValue, { context }) => {
      const { repoURL, commitSHA } = context;
      const { owner, repo } = parseRepoURL(repoURL);
      this.logAPICall('git.getCommit', `${owner} / ${repo} ${commitSHA}`);
      const rootTreeSHA = (
        await this.octokit.rest.git.getCommit({
          owner,
          repo,
          commit_sha: commitSHA,
        })
      ).data.tree.sha;
      this.logAPICall('git.getTree', `${owner} / ${repo} ${rootTreeSHA}`);
      const { tree, truncated } = (
        await this.octokit.rest.git.getTree({
          owner,
          repo,
          tree_sha: rootTreeSHA,
          recursive: 'true',
        })
      ).data;
      const allTreesForCommit: AllTreesForCommit = {
        pathToTreeSHA: new Map(),
        truncated,
      };
      for (const { path, type, sha } of tree) {
        if (
          type === 'tree' &&
          typeof path === 'string' &&
          typeof sha === 'string'
        ) {
          allTreesForCommit.pathToTreeSHA.set(path, sha);
        }
      }
      // Also set the root itself in case we're tracking the root of a repo for
      // some reason.
      allTreesForCommit.pathToTreeSHA.set('', rootTreeSHA);
      return allTreesForCommit;
    },
  });

  async resolveRefToSHA({
    repoURL,
    ref,
  }: ResolveRefToSHAOptions): Promise<string> {
    // If the ref already looks like a SHA, just return it.
    // This does not validate that the commit actually exists in the repo,
    // but in practice the next thing we're going to do is call getTreeSHAForPath
    // with the SHA and that will apply that validation.
    if (ref.match(/^[0-9a-f]{40}$/)) {
      return ref;
    }
    const { owner, repo } = parseRepoURL(repoURL);
    const prNumber = ref.match(/^pr-([0-9]+)$/)?.[1];
    const refParameter = prNumber ? `pull / ${prNumber} / head` : ref;
    this.logAPICall('repos.getCommit', `${owner} / ${repo} ${refParameter}`);
    const sha = (
      await this.octokit.rest.repos.getCommit({
        owner,
        repo,
        ref: refParameter,
        mediaType: {
          format: 'sha',
        },
      })
    ).data as unknown;
    // The TS types don't understand that `mediaType: { format: 'sha' }` turns
    // `.data` into a string, so we have to cast to `unknown` and check
    // ourselves.
    if (typeof sha !== 'string') {
      throw Error('Expected string response');
    }
    return sha;
  }

  async getTreeSHAForPath({
    repoURL,
    commitSHA,
    path,
  }: GetTreeSHAForPathOptions): Promise<string | null> {
    const allTreesForCommit = await this.allTreesForCommitCache.fetch(
      JSON.stringify({ repoURL, commitSHA }),
      { context: { repoURL, commitSHA } },
    );
    if (!allTreesForCommit) {
      // This shouldn't happen: errors should lead to an error being thrown from
      // the previous line, but the fetchMethod always returns an actual item.
      throw Error(`Unexpected missing entry in allTreesForCommitCache`);
    }
    const shaFromCache = allTreesForCommit.pathToTreeSHA.get(path);
    if (shaFromCache) {
      return shaFromCache;
    }
    if (!allTreesForCommit.truncated) {
      // The recursive listing we got from GitHub is complete, so if it doesn't
      // have the tree in question, then the path just doesn't exist (as a tree)
      // at the given commit.
      return null;
    }
    // Hmm, we haven't heard of this tree but our listing was truncated. Fall
    // back to the one-at-a-time API.
    return this.getTreeSHAForPathViaGetContent({ repoURL, commitSHA, path });
  }

  // Fall back to asking the GitHub API for the tree hash directly if our cache
  // was truncated.
  private async getTreeSHAForPathViaGetContent({
    repoURL,
    commitSHA,
    path,
  }: GetTreeSHAForPathOptions): Promise<string | null> {
    const { owner, repo } = parseRepoURL(repoURL);
    this.logAPICall('repos.getContent', `${owner} / ${repo} ${commitSHA}`);
    let data;
    try {
      data = (
        await this.octokit.rest.repos.getContent({
          owner,
          repo,
          ref: commitSHA,
          path,
          mediaType: {
            format: 'object',
          },
        })
      ).data as unknown;
    } catch (e: unknown) {
      // If it looks like "not found" just return null.
      if (typeof e === 'object' && e && 'status' in e && e.status === 404) {
        return null;
      }
      throw e;
    }
    // TS types seem confused here too; this works in practice.
    if (
      !(
        typeof data === 'object' &&
        data !== null &&
        'type' in data &&
        data.type === 'dir' &&
        'sha' in data &&
        typeof data.sha === 'string'
      )
    ) {
      throw Error('response does not appear to be a tree');
    }
    return data.sha;
  }
}

export class CachingGitHubClient {
  constructor(
    private wrapped: GitHubClient,
    dump?: CachingGitHubClientDump,
  ) {
    if (dump) {
      this.getTreeSHAForPathCache.load(dump.treeSHAs);
    }
  }

  private resolveRefToSHACache = new LRUCache<
    string,
    string,
    ResolveRefToSHAOptions
  >({
    max: 1024,
    fetchMethod: async (_key, _staleValue, { context }) => {
      return this.wrapped.resolveRefToSHA(context);
    },
  });

  private getTreeSHAForPathCache = new LRUCache<
    string,
    // LRUCache can't store null, so we box it.
    { boxed: string | null },
    GetTreeSHAForPathOptions
  >({
    max: 1024,
    fetchMethod: async (_key, _staleValue, { context }) => {
      return { boxed: await this.wrapped.getTreeSHAForPath(context) };
    },
  });

  async resolveRefToSHA(options: ResolveRefToSHAOptions): Promise<string> {
    const sha = await this.resolveRefToSHACache.fetch(JSON.stringify(options), {
      context: options,
    });
    if (!sha) {
      throw Error(
        'resolveRefToSHACache.fetch should never resolve without a real SHA',
      );
    }
    return sha;
  }

  async getTreeSHAForPath(
    options: GetTreeSHAForPathOptions,
  ): Promise<string | null> {
    const cached = await this.getTreeSHAForPathCache.fetch(
      JSON.stringify(options),
      {
        context: options,
      },
    );
    if (!cached) {
      throw Error(
        'getTreeSHAForPathCache.fetch should never resolve without a boxed value',
      );
    }
    return cached.boxed;
  }

  async compareCommits(
    options: CompareCommitsOptions,
  ): Promise<CompareCommitsResult | null> {
    return this.wrapped.compareCommits(options);
  }

  dump(): CachingGitHubClientDump {
    // We don't dump resolveRefToSHACache because it is not immutable (it tracks
    // the current commits on main, etc).
    return { treeSHAs: this.getTreeSHAForPathCache.dump() };
  }
}

export interface CachingGitHubClientDump {
  treeSHAs: [
    string,
    LRUCache.Entry<{
      boxed: string | null;
    }>,
  ][];
}

export function isCachingGitHubClientDump(
  dump: unknown,
): dump is CachingGitHubClientDump {
  if (!dump || typeof dump !== 'object') {
    return false;
  }
  if (!('treeSHAs' in dump)) {
    return false;
  }
  const { treeSHAs } = dump;
  if (!Array.isArray(treeSHAs)) {
    return false;
  }

  // XXX we could check the values further if we want to be more anal
  return true;
}
