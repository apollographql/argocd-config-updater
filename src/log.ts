import * as core from '@actions/core';

export class PrefixingLogger {
  private silent = false;

  constructor(private prefix = '') {}

  static silent(): PrefixingLogger {
    const logger = new PrefixingLogger();
    logger.silent = true;
    return logger;
  }

  info(message: string): void {
    if (!this.silent) {
      core.info(this.prefix + message);
    }
  }
  warning(message: string): void {
    if (!this.silent) {
      core.warning(this.prefix + message);
    }
  }
  error(message: string): void {
    if (!this.silent) {
      core.error(this.prefix + message);
    }
  }

  withExtendedPrefix(extension: string): PrefixingLogger {
    const logger = new PrefixingLogger(this.prefix + extension);
    logger.silent = this.silent;
    return logger;
  }
}
