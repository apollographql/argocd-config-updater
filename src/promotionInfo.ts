import { Link } from './templates';

export interface PromotionInfoCommits {
  type: 'commits';
  commitSHAs: string[]; // non-empty
}

export function promotionInfoCommits(
  commitSHAs: string[],
): PromotionInfoCommits {
  return { type: 'commits', commitSHAs };
}

/** Used when the value is being changed but there are no commits between the two
 *  values (ie, no-op change). */
export interface PromotionInfoNoCommits {
  type: 'no-commits';
}

/** Used when the tag/ref value isn't changed at all. */
export interface PromotionInfoNoChange {
  type: 'no-change';
}

// If we're unable to tell what happened with a promotion, the message will
// explain that.
export interface PromotionInfoUnknown {
  type: 'unknown';
  message: string;
}

export function promotionInfoUnknown(message: string): PromotionInfoUnknown {
  return { type: 'unknown', message };
}

export type PromotionInfo =
  | PromotionInfoCommits
  | PromotionInfoNoCommits
  | PromotionInfoNoChange
  | PromotionInfoUnknown;

export interface EnvironmentPromotions {
  environment: string;
  trimmedRepoURL: string;
  gitConfigPromotionInfo: PromotionInfo;
  dockerImage: {
    repository: string;
    promotionInfo: PromotionInfo;
  } | null; // null if there's no Docker image being tracked
  links: Link[];
}
// Map from application-values filenames to EnvironmentPromotions.
export type PromotionsByFile = Map<string, EnvironmentPromotions[]>;
