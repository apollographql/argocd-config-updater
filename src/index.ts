import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import { throttling } from '@octokit/plugin-throttling';
import { eachLimit } from 'async';
import { readFile, writeFile } from 'fs/promises';
import {
  ArtifactRegistryDockerRegistryClient,
  CachingDockerRegistryClient,
  DockerRegistryClient,
} from './artifactRegistry';
import {
  CachingGitHubClient,
  CachingGitHubClientDump,
  GitHubClient,
  OctokitGitHubClient,
  isCachingGitHubClientDump,
} from './github';
import { updateDockerTags } from './update-docker-tags';
import { updateGitRefs } from './update-git-refs';
import { updatePromotedValues } from './update-promoted-values';
import { PrefixingLogger } from './log';
import { inspect } from 'util';

/**
 * The main function for the action.
 * @returns {Promise<void>} Resolves when the action is complete.
 */
export async function main(): Promise<void> {
  try {
    const files = core.getInput('files');
    const globber = await glob.create(files);
    const filenames = await globber.glob();

    core.setOutput(
      'suggested-promotion-branch-name',
      `${core.getInput('promotion-target-regexp')}_${core.getInput('files')}`.replaceAll(
        /[^-a-zA-Z0-9._]/g,
        '_',
      ),
    );

    let gitHubClient: GitHubClient | null = null;
    let finalizeGitHubClient: (() => Promise<void>) | null = null;
    if (core.getBooleanInput('update-git-refs')) {
      const githubToken = core.getInput('github-token');
      const octokit = github.getOctokit(
        githubToken,
        {
          throttle: {
            onRateLimit: (retryAfter, options) => {
              core.warning(
                `[RATE LIMIT] Hit GH rate limit for request ${options.method} ${options.url}; retrying after ${retryAfter} seconds`,
              );
              return true;
            },
            onSecondaryRateLimit: (retryAfter, options) => {
              core.warning(
                `[RATE LIMIT] Hit secondary GH rate limit for request ${options.method} ${options.url}; retrying after ${retryAfter} seconds`,
              );
              return true;
            },
          },
        },
        throttling,
      );

      // Log GH rate limit response headers after each response and at the end.
      let lastRateLimitHeaderInfo: string | null = null;
      octokit.hook.after('request', async (response) => {
        const prefix = 'x-ratelimit-';
        const rateLimitHeaders: string[] = [];
        for (const [name, value] of Object.entries(response.headers)) {
          if (name.startsWith(prefix)) {
            rateLimitHeaders.push(`${name.substring(prefix.length)}=${value}`);
          }
        }

        if (rateLimitHeaders.length) {
          const rateLimitHeaderInfo = rateLimitHeaders.join(', ');
          lastRateLimitHeaderInfo = rateLimitHeaderInfo;
          core.info(`[GH Rate Limit Info] ${rateLimitHeaderInfo}`);
        }
      });

      const octokitGitHubClient = new OctokitGitHubClient(octokit);

      const apiCacheFileName = core.getInput('api-cache');
      let initialAPICache: APICache | null = null;
      if (apiCacheFileName) {
        initialAPICache = await maybeReadAPICache(apiCacheFileName);
      }

      const cachingGitHubClient = new CachingGitHubClient(
        octokitGitHubClient,
        initialAPICache?.gitHub,
      );

      gitHubClient = cachingGitHubClient;

      finalizeGitHubClient = async () => {
        for (const [name, count] of octokitGitHubClient.apiCalls) {
          core.info(`Total GH API calls for ${name}: ${count}`);
        }
        if (lastRateLimitHeaderInfo) {
          core.info(`Last GH Rate Limit Info: ${lastRateLimitHeaderInfo}`);
        }
        if (apiCacheFileName) {
          const finalAPICache: APICache = {
            version: 1,
            gitHub: cachingGitHubClient.dump(),
          };
          await writeFile(apiCacheFileName, JSON.stringify(finalAPICache));
        }
      };
    }

    let dockerRegistryClient: DockerRegistryClient | null = null;
    const artifactRegistryRepository = core.getInput(
      'update-docker-tags-for-artifact-registry-repository',
    );
    if (artifactRegistryRepository) {
      dockerRegistryClient = new CachingDockerRegistryClient(
        new ArtifactRegistryDockerRegistryClient(artifactRegistryRepository),
      );
    }

    const parallelism = +core.getInput('parallelism');
    const errors: { filename: string; error: unknown }[] = [];
    await eachLimit(filenames, parallelism, async (filename) => {
      try {
        await processFile(filename, gitHubClient, dockerRegistryClient);
      } catch (error) {
        errors.push({ filename, error });
      }
    });
    if (errors.length) {
      core.setFailed(
        `Errors occurred while processing ${errors.length} file${errors.length > 1 ? 's' : ''}`,
      );
      for (const { filename, error } of errors) {
        core.error(`Error while processing ${filename}: ${inspect(error)}`);
      }
    } else {
      await finalizeGitHubClient?.();
    }
  } catch (error) {
    // Fail the workflow run if an error occurs
    if (error instanceof Error) core.setFailed(error.message);
  }
}

async function processFile(
  filename: string,
  gitHubClient: GitHubClient | null,
  dockerRegistryClient: DockerRegistryClient | null,
): Promise<void> {
  const shortFilename = filename.startsWith(`${process.cwd()}/`)
    ? filename.substring(process.cwd().length + 1)
    : filename;
  const logger = new PrefixingLogger(`[${shortFilename}] `);
  let contents = await readFile(filename, 'utf-8');

  if (dockerRegistryClient) {
    contents = await updateDockerTags(contents, dockerRegistryClient, logger);
  }

  // The git refs depend on the docker tag potentially so we want to update it after the
  // docker tags are updated.
  if (gitHubClient) {
    contents = await updateGitRefs(contents, gitHubClient, logger);
  }

  if (core.getBooleanInput('update-promoted-values')) {
    const promotionTargetRegexp = core.getInput('promotion-target-regexp');
    contents = await updatePromotedValues(
      contents,
      promotionTargetRegexp || null,
      logger,
    );
  }

  await writeFile(filename, contents);
}

interface APICache {
  version: 1;
  gitHub: CachingGitHubClientDump;
}

async function maybeReadAPICache(
  apiCacheFileName: string,
): Promise<APICache | null> {
  let apiCacheText: string;
  try {
    apiCacheText = await readFile(apiCacheFileName, 'utf8');
  } catch (e) {
    core.error(`Error reading cache file ${apiCacheFileName}, ignoring: ${e}`);
    return null;
  }

  let parsed: unknown;
  try {
    parsed = JSON.parse(apiCacheText);
  } catch (e) {
    core.error(`Error parsing cache file ${apiCacheFileName}, ignoring: ${e}`);
    return null;
  }

  if (
    !parsed ||
    typeof parsed !== 'object' ||
    !('gitHub' in parsed) ||
    !('version' in parsed) ||
    parsed.version !== 1
  ) {
    core.error(
      `Cache file ${apiCacheFileName} has the wrong structure; ignoring`,
    );
    return null;
  }

  if (!isCachingGitHubClientDump(parsed.gitHub)) {
    core.error(
      `Cache file ${apiCacheFileName} has the wrong structure under 'gitHub'; ignoring`,
    );
    return null;
  }

  return { version: parsed.version, gitHub: parsed.gitHub };
}

main();
