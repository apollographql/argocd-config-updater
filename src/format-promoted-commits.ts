import { PromotionsByTargetEnvironment } from './promotionInfo';

export function formatPromotedCommits(
  promotionsByFileThenEnvironment: Map<string, PromotionsByTargetEnvironment>,
): string {
  return [...promotionsByFileThenEnvironment.entries()]
    .sort(([a], [b]) => a.localeCompare(b))
    .map(([filename, promotionsByTargetEnvironment]) => {
      const fileHeader = `* ${filename}\n`;
      const byEnvironment = [...promotionsByTargetEnvironment.entries()]
        .sort(([a], [b]) => a.localeCompare(b))
        .map(([environment, promotionSetWithDockerImage]) => {
          const { promotionSet, dockerImageRepository } =
            promotionSetWithDockerImage;
          const {
            trimmedRepoURL,
            gitConfigPromotionInfo,
            dockerImagePromotionInfo,
            links,
          } = promotionSet;
          const lines = [`  - ${environment}\n`];
          for (const link of links) {
            if (link.url) {
              lines.push(`    + [${link.text}](${link.url})\n`);
            } else {
              lines.push(`    + ${link.text}\n`);
            }
          }
          let alreadyMentionedGitNoCommits = false;
          if (
            dockerImageRepository &&
            dockerImagePromotionInfo &&
            dockerImagePromotionInfo.type !== 'no-change'
          ) {
            let maybeGitConfigNoCommits = '';
            if (gitConfigPromotionInfo.type === 'no-commits') {
              alreadyMentionedGitNoCommits = true;
              maybeGitConfigNoCommits =
                ' (and the git ref for the Helm chart made a no-op change to match)';
            }
            lines.push(
              `    + Changes to Docker image \`${dockerImageRepository}\`${maybeGitConfigNoCommits}\n`,
              ...(dockerImagePromotionInfo.type === 'no-commits'
                ? ['No changes affect the built Docker image.']
                : dockerImagePromotionInfo.type === 'unknown'
                  ? [
                      `Cannot determine set of changes to the Docker image: ${dockerImagePromotionInfo.message}`,
                    ]
                  : dockerImagePromotionInfo.commitSHAs.map(
                      (commitSHA) => `${trimmedRepoURL}/commit/${commitSHA}`,
                    )
              ).map((line) => `      * ${line}\n`),
            );
          }
          if (
            gitConfigPromotionInfo.type !== 'no-change' &&
            !alreadyMentionedGitNoCommits
          ) {
            lines.push(
              `    + Changes to Helm chart\n`,
              ...(gitConfigPromotionInfo.type === 'no-commits'
                ? // This one shows up when the ref changes even though there are no

                  // new commits. This is something we do to try to make the ref
                  // match the Docker tag, so it actually does happen frequently
                  // (though usually only when the Docker tag is making a
                  // substantive change) so this message might end up being a bit
                  // spammy; we can remove it if it's not helpful.
                  [
                    'The git ref for the Helm chart has changed, but there are no new commits in the range.',
                  ]
                : gitConfigPromotionInfo.type === 'unknown'
                  ? [
                      `Cannot determine set of changes to the Helm chart: ${gitConfigPromotionInfo.message}`,
                    ]
                  : gitConfigPromotionInfo.commitSHAs.map(
                      (commitSHA) => `${trimmedRepoURL}/commit/${commitSHA}`,
                    )
              ).map((line) => `      * ${line}\n`),
            );
          }
          return lines.join('');
        });
      return fileHeader + byEnvironment.join('\n');
    })
    .join('');
}
