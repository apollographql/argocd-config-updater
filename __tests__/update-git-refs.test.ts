import { readFile } from 'fs/promises';
import { join } from 'path';
import { GitHubClient } from '../src/github';
import { updateGitRefs } from '../src/update-git-refs';
import { PrefixingLogger } from '../src/log';

const mockGitHubClient: GitHubClient = {
  async resolveRefToSHA({ ref }) {
    if (ref === 'make-it-numeric') {
      return '12345678';
    }
    return `immutable-${ref}-hooray`;
  },
  async getTreeSHAForPath() {
    return `${Math.random()}`;
  },
  async getCommitSHAsForPath() {
    return [];
  },
  async getPullRequest() {
    return { state: 'open' };
  },
};

const logger = PrefixingLogger.silent();

async function fixture(filename: string): Promise<string> {
  return await readFile(
    join(__dirname, '__fixtures__', 'update-git-refs', filename),
    'utf-8',
  );
}

describe('action', () => {
  it('updates git refs', async () => {
    const contents = await fixture('sample.yaml');
    const newContents = await updateGitRefs(
      contents,
      mockGitHubClient,
      new Set<string>(),
      logger,
    );
    expect(newContents).toMatchSnapshot();

    // It should be idempotent in this case.
    expect(
      await updateGitRefs(
        newContents,
        mockGitHubClient,
        new Set<string>(),
        logger,
      ),
    ).toBe(newContents);

    // Update the first one again but freeze one environment.
    expect(
      await updateGitRefs(
        contents,
        mockGitHubClient,
        new Set<string>(['some-service-dev1']),
        logger,
      ),
    ).toMatchSnapshot();
  });

  it('updates git ref when repoURL/path are in `global`', async () => {
    const contents = await fixture('global.yaml');
    expect(
      await updateGitRefs(
        contents,
        mockGitHubClient,
        new Set<string>(),
        logger,
      ),
    ).toMatchSnapshot();
  });

  it('only changes ref when tree sha changes', async () => {
    let treeSHAForNew = 'aaaa';
    const mockGithubClientTreeSHA: GitHubClient = {
      async resolveRefToSHA({ ref }) {
        return ref === 'main' ? 'new' : 'old';
      },
      async getTreeSHAForPath({ commitSHA }) {
        return commitSHA === 'old' ? 'aaaa' : treeSHAForNew;
      },
      async getCommitSHAsForPath() {
        return [];
      },
      async getPullRequest() {
        return { state: 'open' };
      },
    };

    const contents = await fixture('tree-sha.yaml');

    // First snapshot: ref should still be 'old' because tree SHA matches.
    expect(
      await updateGitRefs(
        contents,
        mockGithubClientTreeSHA,
        new Set<string>(),
        logger,
      ),
    ).toMatchSnapshot();

    treeSHAForNew = 'bbbb';
    // Second snapshot: ref should now be 'new' because tree SHA has changed.
    expect(
      await updateGitRefs(
        contents,
        mockGithubClientTreeSHA,
        new Set<string>(),
        logger,
      ),
    ).toMatchSnapshot();
  });

  it('handles unknown current ref', async () => {
    const mockGithubClientTreeSHA: GitHubClient = {
      async resolveRefToSHA({ ref }) {
        if (ref === 'old') {
          throw Error('unknown ref');
        }
        return 'new';
      },
      async getTreeSHAForPath() {
        return 'aaaa';
      },
      async getCommitSHAsForPath() {
        return [];
      },
      async getPullRequest() {
        return { state: 'open' };
      },
    };

    const contents = await fixture('tree-sha.yaml');

    expect(
      await updateGitRefs(
        contents,
        mockGithubClientTreeSHA,
        new Set<string>(),
        logger,
      ),
    ).toMatchSnapshot();
  });

  it('handle docker tag', async () => {
    const mockGithubClientTreeSHA: GitHubClient = {
      async resolveRefToSHA({ ref }) {
        return ref === 'main' ? 'new' : ref;
      },
      async getTreeSHAForPath({ commitSHA }) {
        return commitSHA === 'd97b3a3240' ? 'bad' : 'aaaa';
      },
      async getCommitSHAsForPath() {
        return [];
      },
      async getPullRequest() {
        return { state: 'open' };
      },
    };

    const contents = await fixture('docker-tree-sha.yaml');

    expect(
      await updateGitRefs(
        contents,
        mockGithubClientTreeSHA,
        new Set<string>(),
        logger,
      ),
    ).toMatchSnapshot();
  });

  it('handle docker tag with unknown commit', async () => {
    const mockGithubClientTreeSHA: GitHubClient = {
      async resolveRefToSHA({ ref }) {
        if (ref === 'd97b3a3240') {
          throw Error('unknown commit');
        }
        return ref === 'main' ? 'new' : ref;
      },
      async getTreeSHAForPath({ commitSHA }) {
        return commitSHA === 'old' ? 'oldaaaa' : 'aaaa';
      },
      async getCommitSHAsForPath() {
        return [];
      },
      async getPullRequest() {
        return { state: 'open' };
      },
    };

    const contents = await fixture('docker-tree-sha.yaml');

    expect(
      await updateGitRefs(
        contents,
        mockGithubClientTreeSHA,
        new Set<string>(),
        logger,
      ),
    ).toMatchSnapshot();
  });
});
