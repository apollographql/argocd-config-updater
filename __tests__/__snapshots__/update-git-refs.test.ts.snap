// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action handle docker tag 1`] = `
"updated:
  track: main
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: d97b3a3241
  dockerImage:
    tag: main---0013323-2024.03-gd97b3a3241


no-sha-found:
  gitConfig:
    trackMutableRef: main
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: old
  dockerImage:
    tag: main---0013323-2024.03-d97b3a3

diff-tree-sha:
  track: main
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: old
  dockerImage:
    tag: main---0013323-2024.03-gd97b3a3240"
`;

exports[`action handle docker tag with unknown commit 1`] = `
"updated:
  track: main
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: d97b3a3241
  dockerImage:
    tag: main---0013323-2024.03-gd97b3a3241


no-sha-found:
  gitConfig:
    trackMutableRef: main
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: new
  dockerImage:
    tag: main---0013323-2024.03-d97b3a3

diff-tree-sha:
  track: main
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: new
  dockerImage:
    tag: main---0013323-2024.03-gd97b3a3240"
`;

exports[`action handles unknown current ref 1`] = `
"some-service-tree:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    trackMutableRef: main
    ref: new
"
`;

exports[`action only changes ref when tree sha changes 1`] = `
"some-service-tree:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    trackMutableRef: main
    ref: old
"
`;

exports[`action only changes ref when tree sha changes 2`] = `
"some-service-tree:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    trackMutableRef: main
    ref: new
"
`;

exports[`action updates git ref when repoURL/path are in \`global\` 1`] = `
"global:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world

some-service-dev1:
  gitConfig:
    ref:      'immutable-pr-1234-hooray'   # this should change but stay single-quoted!
    trackMutableRef: pr-1234
"
`;

exports[`action updates git refs 1`] = `
"some-service-dev0:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: "immutable-pr-1234-hooray"   # this should change but stay double-quoted!
    trackMutableRef: pr-1234

some-service-dev1:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref:      'immutable-pr-1234-hooray'   # this should change but stay single-quoted!
    trackMutableRef: pr-1234

some-service-staging:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    trackMutableRef: main

    ref:        immutable-main-hooray  #this can stay unquoted because it won't look like a number

some-service-prod:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: abcdef1234    # NO TOUCHING! No tracking.

another:
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: '12345678'   # this becomes a number and will definitely need to be quoted
    trackMutableRef: make-it-numeric

some-service-top-level:
  track: pr-1234
  gitConfig:
    repoURL: https://github.com/apollographql/some-repo.git
    path: services/hello-world
    ref: "immutable-pr-1234-hooray"   # this should change but stay double-quoted!
"
`;
